x-pack/legacy/plugins/reporting/server/browsers/network_policy.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import * as _ from 'lodash';
+import * as url from 'url';
+
+interface FirewallRequest {
+  ip: string | null;
+  url: string;
+}
+
+interface FirewallRule {
+  allow: boolean;
+  hosts?: string[];
+  protocols?: string[];
+  ips?: string[];
+}
+
+export const allowRequest = (request: FirewallRequest, rules: FirewallRule[]) => {
+  const parsed = url.parse(request.url);
+
+  if (!rules.length) {
+    return true;
+  }
+
+  // Accumulator has three potential values here:
+  // True => allow request, don't check other rules
+  // False => reject request, don't check other rules
+  // Undefined => Not yet known, proceed to next rule
+  const allowed = rules.reduce((result: boolean | undefined, rule) => {
+    if (typeof result === 'boolean') {
```

Comment:
This coupled with `return rule.allow || isRuleMatch ? rule.allow && isRuleMatch : undefined;` creates some interesting  behaviors. Some of which I'm unsure how I feel about, and one that I feel is wrong...

When we have a single `allow` rule, all other other requests will be denied. This in and of itself seems fine.
```
  it('when no rules match, but we have a single allow', () => {
    const rules = [
      { allow: true, hosts: ['foo.com']},
    ];

    expect(allowRequest({ url: 'https://bar.com', ip: null }, rules)).toBe(false);
  });
```

However, if we have a single deny statement all other requests will be allowed. This generally makes sense, but we'll have to clearly and explicitly document the behavior as it's somewhat different than the above.
```
  it('when no rules match, but we have a single deny', () => {
    const rules = [
      { allow: false, hosts: ['bar.com']},
    ];

    expect(allowRequest({ url: 'https://quz.com', ip: null }, rules)).toBe(true);
  });
```

The following just seems wrong though:
```
  it('allow, deny, allow', () => {
    const rules = [
      { allow: true, hosts: ['bar.com']},
      { allow: false, hosts: ['baz.com']},
      { allow: true, hosts: ['foo.com']},
    ];

    expect(allowRequest({ url: 'https://foo.com', ip: null }, rules)).toBe(true);
  });
```

I very well might be wrong about this, but it feels like we should be denying all requests which don't match a specific rule, and allowing the user to specify a rule that applies to all requests. This way, worst case scenario we end up denying a request (which is safer than allowing one). And we explicitly require the end-user to specify a rule stating that all other requests that haven't matches one of the other rules are allowed.